base/pointer.jl

```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 
 """
     unsafe_load(p::Ptr{T}, i::Integer=1)
+    unsafe_load(p::Ptr{T}, order::Symbol)
+    unsafe_load(p::Ptr{T}, i::Integer, order::Symbol)
 
 Load a value of type `T` from the address of the `i`th element (1-indexed) starting at `p`.
-This is equivalent to the C expression `p[i-1]`.
+This is equivalent to the C expression `p[i-1]`. Optionally, an ordering can be provided.
 
 The `unsafe` prefix on this function indicates that no validation is performed on the
 pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
 that referenced memory is not freed or garbage collected while invoking this function.
 Incorrect usage may segfault your program or return garbage answers. Unlike C, dereferencing
 memory region allocated as different type may be valid provided that the types are compatible.
+
+!!! compat "Julia 1.10"
+     The `order` argument is available as of Julia 1.10.
 """
 unsafe_load(p::Ptr, i::Integer=1) = pointerref(p, Int(i), 1)
+unsafe_load(p::Ptr, order::Symbol) = atomic_pointerref(p, order)
+function unsafe_load(p::Ptr, i::Integer, order::Symbol)
+    unsafe_load(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), order)
+end
 
 """
     unsafe_store!(p::Ptr{T}, x, i::Integer=1)
+    unsafe_store!(p::Ptr{T}, x, order::Symbol)
+    unsafe_store!(p::Ptr{T}, x, i::Integer, order::Symbol)
 
 Store a value of type `T` to the address of the `i`th element (1-indexed) starting at `p`.
-This is equivalent to the C expression `p[i-1] = x`.
+This is equivalent to the C expression `p[i-1] = x`. Optionally, an ordering can be
+provided.
 
 The `unsafe` prefix on this function indicates that no validation is performed on the
 pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
 that referenced memory is not freed or garbage collected while invoking this function.
 Incorrect usage may segfault your program. Unlike C, storing memory region allocated as
 different type may be valid provided that that the types are compatible.
+
+!!! compat "Julia 1.10"
+     The `order` argument is available as of Julia 1.10.
 """
 unsafe_store!(p::Ptr{Any}, @nospecialize(x), i::Integer=1) = pointerset(p, x, Int(i), 1)
 unsafe_store!(p::Ptr{T}, x, i::Integer=1) where {T} = pointerset(p, convert(T,x), Int(i), 1)
+unsafe_store!(p::Ptr, x, order::Symbol) = atomic_pointerset(p, x, order)
+function unsafe_store!(p::Ptr, x, i::Integer, order::Symbol)
+    unsafe_store!(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), x, order)
+end
+
+"""
+    unsafe_modify!(p::Ptr{T}, op, x, [order::Symbol]) -> Pair
+
+These atomically perform the operations to get and set a field after applying
```

Comment:
nit: it is not really legal to use this on a field, since those have slightly different semantics, so it should only really be for interop with existing C libraries that require it. It may be helpful also to add an additional statement that this is compatible with C11 and C++11 memory semantics for `_Atomic` and `std::atomic` respectively, per the C++23 standard.
```suggestion
These atomically perform the operations to get and set a memory address after applying
```

The relevant C++23 standard text wording that would be applicable here is:
```
Implementations are recommended to ensure that the representation of
_Atomic(T) in C is same as that of std::atomic<T> in C++ for every
possible type T. The mechanisms used to ensure atomicity and memory
ordering should be compatible.
```

Reply:
How about this boiler plate

```
The `unsafe` prefix on this function indicates that no validation is performed on the
pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
that referenced memory is not freed or garbage collected while invoking this function.
Incorrect usage may segfault your program. Unlike C, storing memory region allocated as
different type may be valid provided that that the types are compatible.

This is compatible with C11 and C++11 memory semantics for `_Atomic` and `std::atomic`
respectively, per the C++23 standard. Therefore, implementations that atomically
manipulate `Ptr{T}` are recommended to ensure that representation of `_Atomic(T)` in C.
The mechanisms used to ensure atomicity and memory ordering should be compatible.

```

Reply:
seems like this phrase lots some of the grammatical helper words when I originally wrote it:
"...storing to a memory region allocated as a different type is valid, provided..."
(or `accessing`, or `reading from`, depending on context)

Reply:
> Therefore, implementations that atomically
manipulate `Ptr{T}` are recommended to ensure that representation of `_Atomic(T)` in C.
The mechanisms used to ensure atomicity and memory ordering should be compatible.

There is only one Julia implementation, so we aren't writing a spec for implementations of it. This should convey that the Julia implementation is compatible with use on pointers declared with `_Atomic` and `std::atomic` type in C11 and C++23 respectively. But not all C types compatible with `_Atomic` are translated at present to work in Julia, so the operation may instead throw a Julia error that the type `T` is not supported.

Reply:
> There is only one Julia implementation, so we aren't writing a spec for implementations of it.

Sorry, poor use of words on my part. I was trying to find an more elegant way of saying "code that passes `Ptr{T}` to a method with a atomic order argument should ensure ..."

> This should convey that the Julia implementation is compatible with use on pointers declared with _Atomic and std::atomic type in C11 and C++23 respectively. But not all C types compatible with _Atomic are translated at present to work in Julia, so the operation may instead throw a Julia error that the type T is not supported.

I wonder if some of this should start going in "doc/src/base/multi-threading.md" under the "Atomic Operations" heading since its relevant to atomic operations in Julia as a whole and not just these methods. It's also starting to be quite a big chunk of text if we want a light explanation on what these do in the REPL

Reply:
Yeah, it is probably best to put it in multi-threading.md and cross-reference it. We could also potentially make a kwdoc for it too, so it can be queried from the REPL help (in base/docs/basedocs.jl)

```julia
"""
For unsafe pointer operations, ...
And see also...
"""
kw"atomic"
```